src/wasm/wasm-type-shape.cpp

```diff
@@ -0,0 +1,348 @@
+      return aSuper.has_value() < bSuper.has_value() ? LT : GT;
+    }
+    if (aSuper) {
+      if (auto cmp = compare(*aSuper, *bSuper); cmp != EQ) {
+        return cmp;
+      }
+    }
+    auto aKind = a.getKind();
+    auto bKind = b.getKind();
+    if (aKind != bKind) {
+      return aKind < bKind ? LT : GT;
+    }
+    switch (aKind) {
+      case HeapTypeKind::Func:
+        return compare(a.getSignature(), b.getSignature());
+      case HeapTypeKind::Struct:
+        return compare(a.getStruct(), b.getStruct());
+      case HeapTypeKind::Array:
+        return compare(a.getArray(), b.getArray());
+      case HeapTypeKind::Cont:
+        return compare(a.getContinuation(), b.getContinuation());
+      case HeapTypeKind::Basic:
+        break;
+    }
+    WASM_UNREACHABLE("unexpected kind");
+  }
+
+  Comparison compare(Signature a, Signature b) {
+    if (auto cmp = compare(a.params, b.params); cmp != EQ) {
+      return cmp;
+    }
+    return compare(a.results, b.results);
+  }
+
+  Comparison compare(const Struct& a, const Struct& b) {
+    if (a.fields.size() != b.fields.size()) {
+      return a.fields.size() < b.fields.size() ? LT : GT;
+    }
+    for (size_t i = 0; i < a.fields.size(); ++i) {
+      if (auto cmp = compare(a.fields[i], b.fields[i]); cmp != EQ) {
+        return cmp;
+      }
+    }
+    return EQ;
+  }
+
+  Comparison compare(Array a, Array b) { return compare(a.element, b.element); }
+
+  Comparison compare(Continuation a, Continuation b) {
+    return compare(a.type, b.type);
+  }
+
+  Comparison compare(Field a, Field b) {
+    if (a.mutable_ != b.mutable_) {
+      return a.mutable_ < b.mutable_ ? LT : GT;
+    }
+    if (a.isPacked() != b.isPacked()) {
+      return b.isPacked() < a.isPacked() ? LT : GT;
+    }
+    if (a.packedType != b.packedType) {
+      return a.packedType < b.packedType ? LT : GT;
+    }
+    return compare(a.type, b.type);
+  }
+
+  Comparison compare(Type a, Type b) {
+    if (a.isBasic() != b.isBasic()) {
+      return b.isBasic() < a.isBasic() ? LT : GT;
+    }
+    if (a.isBasic()) {
+      if (a.getBasic() != b.getBasic()) {
+        return a.getBasic() < b.getBasic() ? LT : GT;
+      }
+      return EQ;
+    }
+    if (a.isTuple() != b.isTuple()) {
+      return a.isTuple() < b.isTuple() ? LT : GT;
+    }
+    if (a.isTuple()) {
+      return compare(a.getTuple(), b.getTuple());
+    }
+    assert(a.isRef() && b.isRef());
+    if (a.isNullable() != b.isNullable()) {
+      return a.isNullable() < b.isNullable() ? LT : GT;
+    }
+    return compare(a.getHeapType(), b.getHeapType());
+  }
+
+  Comparison compare(const Tuple& a, const Tuple& b) {
+    if (a.size() != b.size()) {
+      return a.size() < b.size() ? LT : GT;
+    }
+    for (size_t i = 0; i < a.size(); ++i) {
+      if (auto cmp = compare(a[i], b[i]); cmp != EQ) {
+        return cmp;
+      }
+    }
+    return EQ;
+  }
+
+  Comparison compare(HeapType a, HeapType b) {
```

Comment:
What is the difference between `compareDefinition(HeapType, HeapType)` and `compare(HeapType, HeapType)`?

Reply:
`compareDefinition` compares the structures of two heap types, whereas `compare` compares the heap types without recursing into their structures, for example by seeing if they have the same index in their respective recursion groups or if they are actually identical.

Reply:
What does "actually identical" mean in the last sentence, if not to look at their structure? Is that the external comparator I see called at the end?

Reply:
Right, it calls out to some user-provided method for ordering heap types. Or for the case of equality, it checks that the types are equivalent using `==`.